chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+Reducing the type checking function to what is relevant to the linearization process yields [@lst:nickel-tc-abstract].
+Essentially, every term is unconditionally registered by the linearization.
+This is enough to handle a large subset of Nickel.
+In fact, only records, let bindings and function definitions require additional change to enrich identifiers they define with type information.
+
+
+```{.rust #lst:nickel-tc-abstract caption="Abstract type checking function"}
+fn type_check_<L: Linearizer>(
+    lin: &mut Linearization<L::Building>,
+    mut linearizer: L,
+    rt: &RichTerm,
+    ty: TypeWrapper,
+    /* omitted */
+) -> Result<(), TypecheckError> {
+    let RichTerm { term: t, pos } = rt;
+
+    // 1. record a node
+    linearizer.add_term(lin, t, *pos, ty.clone());
+
+    // handling of each term variant
+    // recursively calling `type_check_`
+    //
+    // 2. retype identifiers if needed
+    match t.as_ref() {
+      Term::RecRecord(stat_map, ..) => {
+        for (id, rt) in stat_map {
+          let tyw = binding_type(/* omitted */);
+          linearizer.retype_ident(lin, id, tyw);
+        }
+      }
+      Term::Fun(ident, _) |
+      Term::FunPattern(Some(ident), _)=> {
+        let src = state.table.fresh_unif_var();
+        linearizer.retype_ident(lin, ident, src.clone());
+      }
+      Term::Let(ident, ..) |
+      Term::LetPattern(Some(ident), ..)=> {
+        let ty_let = binding_type(/* omitted */);
+        linearizer.retype_ident(lin, ident, ty_let.clone());
+      }
+      _ => { /* omitted */ }
+    }
+```
+
+While registering a node, NLS distinguishes 4 kinds of nodes.
+These are *metadata*, *usage graph* related nodes, i.e. declarations and usages, *static access* of nested record fields, and *general elements* which is every node that does not fall into one of the prior categories.
+
+
+##### Structures
+
+```{.nickel #lst:nickel-simple-expr caption="Exemplary nickel expressions"}
+// atoms
+
+1
+true
+null
+
+// binary operations
+42 * 3
+[ 1, 2, 3 ] @ [ 4, 5]
+
+// if-then-else
+if true then "TRUE :)" else "false :("
+
+// string iterpolation
+"#{ "hello" } #{ "world" }!"
+```
+
+In the most common case of general elements, the node is simply registered as a `LinearizationItem` of kind `Structure`.
+This applies for all simple expressions like those exemplified in [@lst:nickel-simple-expr]
+
+
+##### Declarations
+
+In case of `let` bindings or function arguments name binding is equally simple.
+As discussed in [@sec:let-bindings-and-functions] the `let` node may contain both a name and pattern matches.
+For either the linearizer generates `Declaration` items and updates its name register.
+However, type information is available for name bindings only, meaning pattern matches remain untyped.
+
+The same process applies for argument names in function declarations.
+Due to argument currying, NLS linearizes only a single argument/pattern at a time.
```

Comment:
Suggestion: maybe add a link to a general entry about currying here, which is a well-known notion but still not for anyone outside of functional programming  (I guess the wikipedia page would do the job)